magento/magento2#37983: Place order with disabled Payment method working

KrasnoshchokBohdan opened the pull request:
- prevent order placement with unavailable payment method

<!---
    Thank you for contributing to Magento.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
I will leave here examples of mutations that I used.
<details>
  <summary>createCustomer</summary>

```graphql
mutation {
  createCustomer(
    input: {
      firstname: "Test"
      lastname: "User"
      email: "test.user@example.com"
      password: "Password123"
    }
  ) {
    customer {
      id
      email
    }
  }
}
```
</details>
<details>
  <summary>generateCustomerToken</summary>

```graphql
mutation {
  generateCustomerToken(
    email: "test.user@example.com"
    password: "Password123"
  ) {
    token
  }
}
```
</details>
<details>
  <summary>createEmptyCart (don't forget about Authentification)</summary>

```graphql
mutation {
  createEmptyCart
}
```
</details>
<details>
  <summary>addSimpleProductsToCart </summary>

```graphql
mutation {
  addSimpleProductsToCart(
    input: {
      cart_id: "SOMEID" # Replace with the ID from Step 3
      cart_items: [
        {
          data: {
            quantity: 1
            sku: "test2" # Replace with an actual product SKU from your store
          }
        }
      ]
    }
  ) {
    cart {
      items {
        product {
          name
          sku
        }
        quantity
      }
    }
  }
}
```
</details>
<details>
  <summary>setShippingAddressesOnCart</summary>

```graphql
mutation {
  setShippingAddressesOnCart(
    input: {
      cart_id: "SOMEID" # Replace with the ID from Step 3
      shipping_addresses: [
        {
          address: {
            firstname: "Test"
            lastname: "User"
            company: "Company Name"
            street: ["123 Main Street"]
            city: "New York"
            region: "NY"
            postcode: "10001"
            country_code: "US"
            telephone: "1234567890"
            save_in_address_book: false
          }
        }
      ]
    }
  ) {
    cart {
      shipping_addresses {
        firstname
        lastname
        street
        city
        region {
          code
          label
        }
        postcode
        country {
          code
          label
        }
        telephone
      }
    }
  }
}
```
</details>
<details>
  <summary>setBillingAddressOnCart</summary>

```graphql
mutation {
  setBillingAddressOnCart(
    input: {
      cart_id: "SOMEID" # Replace with the ID from Step 3
      billing_address: {
        address: {
          firstname: "Test"
          lastname: "User"
          company: "Company Name"
          street: ["123 Main Street"]
          city: "New York"
          region: "NY"
          postcode: "10001"
          country_code: "US"
          telephone: "1234567890"
          save_in_address_book: false
        }
        same_as_shipping: false
      }
    }
  ) {
    cart {
      billing_address {
        firstname
        lastname
        street
        city
        region {
          code
          label
        }
        postcode
        country {
          code
          label
        }
        telephone
      }
    }
  }
}
```
</details>
<details>
  <summary>setShippingMethodsOnCart</summary>

```graphql
mutation {
  setShippingMethodsOnCart(
    input: {
      cart_id: "SOMEID" # Replace with the ID from Step 3
      shipping_methods: [
        {
          carrier_code: "flatrate"
          method_code: "flatrate"
        }
      ]
    }
  ) {
    cart {
      shipping_addresses {
        selected_shipping_method {
          carrier_code
          method_code
          carrier_title
          method_title
        }
      }
    }
  }
}
```
</details>
<details>
  <summary>setPaymentMethodOnCart + placeOrder</summary>

```graphql
mutation setPaymentMethodAndPlaceOrder(
  $cartId: String!
  $paymentMethod: PaymentMethodInput!
) {
  setPaymentMethodOnCart(
    input: { cart_id: $cartId, payment_method: $paymentMethod }
  ) {
    cart {
      selected_payment_method {
        code
        title
      }
    }
  }

  placeOrder(input: { cart_id: $cartId }) {
    order {
      order_number
    }
  }
}
```
VARIABLES
```graphql
{
    "cartId": "SOMEID",
    "paymentMethod": {
        "code": "checkmo"
    }
}
```
</details>


### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format magento/magento2#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
1. Fixes magento/magento2#[37983](https://github.com/magento/magento2/issues/37983)

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Disable the checkmo payment method in admin
2. Create a quote using GraphQL
3. Set addresses and shipping method in GraphQL
4. Use a combined mutation with setPaymentMethodOnCart and placeOrder
5. Observe that despite the payment method error, an order is still created


### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] README.md files for modified modules are updated and included in the pull request if any [README.md predefined sections](https://github.com/magento/devdocs/wiki/Magento-module-README.md) require an update
 - [ ] All automated tests passed successfully (all builds are green)

KrasnoshchokBohdan committed:
magento/magento2#37983: Place order with disabled Payment method working

- prevent order placement with unavailable payment method

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-37983

KrasnoshchokBohdan committed:
magento/magento2#37983: Place order with disabled Payment method working

- copyright and static test errors fix

KrasnoshchokBohdan committed:
Merge remote-tracking branch 'origin/fix-for-issue-37983' into fix-for-issue-37983

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-37983

KrasnoshchokBohdan commented:
@magento run all tests

KrasnoshchokBohdan committed:
magento/magento2#37983: Place order with disabled Payment method working

Updated the code to safely handle cases where the cart's payment object might be null. This prevents potential errors when accessing the payment method code for carts without a valid payment set.

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-37983

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-37983

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-37983

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-37983

engcom-Hotel commented on app/code/Magento/QuoteGraphQl/Model/Cart/PlaceOrder.php:
```diff
@@ -58,6 +58,26 @@ public function execute(Quote $cart, string $maskedCartId, int $userId): int
+        $availablePaymentMethods = $this->paymentManagement->getList($cartId);
+        $payment = $cart->getPayment();
+        $paymentMethodCode = $payment?->getMethod();
+        $isPaymentMethodAvailable = false;
+
+        // Check if the selected payment method is in the available methods list
+        if ($paymentMethodCode && $availablePaymentMethods) {
+            foreach ($availablePaymentMethods as $availableMethod) {
+                if ($availableMethod->getCode() === $paymentMethodCode) {
+                    $isPaymentMethodAvailable = true;
+                    break;
+                }
+            }
+        }
```
Suggesting the below changes:
```suggestion
        // Check if the selected payment method is in the available methods list
        if ($paymentMethodCode && $availablePaymentMethods) {
            $availableCodes = array_map(fn($method) => $method->getCode(), $availablePaymentMethods);
            $isPaymentMethodAvailable = in_array($paymentMethodCode, $availableCodes);
        } else {
            $isPaymentMethodAvailable = false;
        }
```

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for the contribution!

Please refer to the below review comment and also please add some automated test in accordance to the [DOD](https://developer.adobe.com/commerce/contributor/guides/code-contributions/definition-of-done/).

Thanks

KrasnoshchokBohdan committed:
Update app/code/Magento/QuoteGraphQl/Model/Cart/PlaceOrder.php

Co-authored-by: Abhinav Pathak <51681618+engcom-Hotel@users.noreply.github.com>

KrasnoshchokBohdan committed:
magento/magento2#37983: Place order with disabled Payment method working

- Add unit tests for PlaceOrder functionality in QuoteGraphQl.

KrasnoshchokBohdan commented:
@magento run all tests

engcom-Hotel commented:
Hello @KrasnoshchokBohdan,

It seems that you have made the changes according to the review comments, we are picking this PR for the review.

Thanks

engcom-Hotel commented on app/code/Magento/QuoteGraphQl/Model/Cart/PlaceOrder.php:
```diff
@@ -58,6 +58,22 @@ public function execute(Quote $cart, string $maskedCartId, int $userId): int
+        }
+
+        if (!$isPaymentMethodAvailable) {
+            throw new LocalizedException(__('The requested Payment Method is not available.'));
```
The error message could be more specific by including the requested payment method code: "The requested Payment Method '{code}' is not available."

And also consider adding a debug log when a disabled payment method is detected to help track these attempts.

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for making the changes. Just a minor changes requested below. Please do the needful.

Thanks

KrasnoshchokBohdan committed:
magento/magento2#37983: Place order with disabled Payment method working

Add logging for disabled payment method attempts in PlaceOrder

Extended `PlaceOrder` to log debug information when an unavailable payment method is used. This includes details like the cart ID, user ID, attempted payment method, and available methods, facilitating easier troubleshooting of order placement issues. Updated unit tests to validate the new logging behavior.

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-37983

engcom-Hotel committed:
Merge branch '2.4-develop' into fix-for-issue-37983

engcom-Hotel commented:
@magento run all tests

engcom-Hotel commented on app/code/Magento/QuoteGraphQl/Model/Cart/PlaceOrder.php:
```diff
@@ -28,16 +29,24 @@ class PlaceOrder
         PaymentMethodManagementInterface $paymentManagement,
-        CartManagementInterface $cartManagement
+        CartManagementInterface $cartManagement,
+        LoggerInterface $logger
```
New Params in Constructor should be null and handle it with ObjectManager

engcom-Hotel reviewed (changes requested):
Hello @KrasnoshchokBohdan,

Thanks for updating the PR!

It seems some automated tests are failing due to this PR change. Please fix those also look into the below review comment.

Thanks

KrasnoshchokBohdan committed:
magento/magento2#37983: Place order with disabled Payment method working

Refactor PlaceOrder to support optional logger dependency

Modified the constructor to make LoggerInterface optional by defaulting to ObjectManager resolution. This improves flexibility and ensures compatibility with cases where a logger is not explicitly provided.

KrasnoshchokBohdan commented:
@magento run Integration Tests, Unit Tests, Functional Tests CE

engcom-Hotel reviewed (changes requested):
@KrasnoshchokBohdan Some tests are still failing due to the PR changes. Moving back to `Changes Requested` bucket.

Thanks

KrasnoshchokBohdan committed:
magento/magento2#37983: Place order with disabled Payment method working

- fix web api tests

KrasnoshchokBohdan commented:
@magento run web api tests

KrasnoshchokBohdan committed:
magento/magento2#37983: Place order with disabled Payment method working

Update error messages in PlaceOrderTest to provide a consistent server error response

KrasnoshchokBohdan committed:
Merge branch '2.4-develop' into fix-for-issue-37983
